network/src/primary.rs

```diff
@@ -0,0 +1,217 @@
+use crate::{CancelHandler, RetryConfig};
+use crypto::traits::VerifyingKey;
+use futures::FutureExt;
+use rand::{prelude::SliceRandom as _, rngs::SmallRng, SeedableRng as _};
+use std::{collections::HashMap, net::SocketAddr};
+use tokio::task::JoinHandle;
+use tonic::transport::Channel;
+use types::{
+    BincodeEncodedPayload, PrimaryMessage, PrimaryToPrimaryClient, PrimaryToWorkerClient,
+    PrimaryWorkerMessage,
+};
+
+pub struct PrimaryNetwork {
+    clients: HashMap<SocketAddr, PrimaryToPrimaryClient<Channel>>,
+    retry_config: RetryConfig,
+    /// Small RNG just used to shuffle nodes and randomize connections (not crypto related).
+    rng: SmallRng,
+}
+
+impl Default for PrimaryNetwork {
+    fn default() -> Self {
+        let retry_config = RetryConfig {
+            // Retry for forever
+            retrying_max_elapsed_time: None,
+            ..Default::default()
+        };
+
+        Self {
+            clients: Default::default(),
+            retry_config,
+            rng: SmallRng::from_entropy(),
+        }
+    }
+}
+
+impl PrimaryNetwork {
+    fn client(&mut self, address: SocketAddr) -> PrimaryToPrimaryClient<Channel> {
+        self.clients
+            .entry(address)
+            .or_insert_with(|| Self::create_client(address))
+            .clone()
+    }
+
+    fn create_client(address: SocketAddr) -> PrimaryToPrimaryClient<Channel> {
+        // TODO use TLS
+        let url = format!("http://{}", address);
+        let channel = Channel::from_shared(url)
+            .expect("URI should be valid")
+            .connect_lazy();
+        PrimaryToPrimaryClient::new(channel)
+    }
+
+    pub async fn send<T: VerifyingKey>(
+        &mut self,
+        address: SocketAddr,
+        message: &PrimaryMessage<T>,
+    ) -> CancelHandler<()> {
+        let message =
+            BincodeEncodedPayload::try_from(message).expect("Failed to serialize payload");
+        self.send_message(address, message).await
+    }
+
+    async fn send_message(
+        &mut self,
+        address: SocketAddr,
+        message: BincodeEncodedPayload,
+    ) -> CancelHandler<()> {
+        let client = self.client(address);
+        let handle = tokio::spawn(
+            self.retry_config
+                .retry(move || {
+                    let mut client = client.clone();
+                    let message = message.clone();
+                    async move { client.send_message(message).await.map_err(Into::into) }
+                })
+                .map(|response| {
+                    response.expect("we retry forever so this shouldn't fail");
```

Comment:
In this context, the endpoint is not trusted with all that this implies.
- for replacing instances of the `reliable_sender` it's Ok to use this send-with-retry, but not for instances of the `simple_sender`, which is latency-sensitive,
- the endpoint you create here in `create_client` should really be under a connect timeout, https://github.com/hyperium/tonic/commit/2b60a00614c5c4260ce0acaaa599da89bebfd267